packages/resolver/src/adapters/cns-cardano-api-maker.ts

```diff
@@ -0,0 +1,121 @@
+import {CNSMetadata, CNSUserRecord} from './cns-types'
+import {hexToString} from './cns-utils'
+import {FetchData, fetchData, handleApiError, isLeft} from '@yoroi/common'
+import {AxiosRequestConfig} from 'axios'
+import {z} from 'zod'
+
+export const makeCnsCardanoApi = (
+  baseUrl: string,
+  fetcherConfig?: AxiosRequestConfig,
+  request: FetchData = fetchData,
+) => {
+  const getAssetAddress = async (
+    assetHex: string,
+  ): Promise<string | undefined> => {
+    const policyId = assetHex.slice(0, 56)
+    const assetName = assetHex.slice(56)
+
+    const response = await request<string>(
+      {
+        url: `${baseUrl}/api/asset/accounts?policy=${policyId}&asset=${assetName}`,
+      },
+      fetcherConfig,
+    )
+
+    if (isLeft(response)) {
+      handleApiError(response.error)
+    } else {
+      const parsedResponse = getAssetAddressSchema.parse(response.value.data)
+      return parsedResponse[0]
+    }
+  }
+
+  const getMetadata = async (
+    policyID: string,
+    assetName: string,
+  ): Promise<CNSMetadata | null> => {
+    const assetNameString = hexToString(assetName)
+    const key = `${policyID}.${assetNameString}`
+
+    const response = await request<CNSMetadata | null>(
+      {
+        url: `${baseUrl}/api/multiAsset/metadata`,
+        method: 'post',
+        data: {assets: [{nameHex: assetName, policy: policyID}]},
+      },
+      fetcherConfig,
+    )
+
+    if (isLeft(response)) {
+      handleApiError(response.error)
+    } else {
+      const parsedResponse = getMetadataSchema.parse(response.value.data)
+
+      if (Object.keys(parsedResponse).length === 0) return null
+
+      // @ts-ignore
+      return parsedResponse[key][0].metadata[policyID][assetNameString] ?? null
+    }
+  }
+
+  const getAssetInlineDatum = async (
+    assetHex: string,
+    addresses: Array<string>,
+  ): Promise<CNSUserRecord> => {
+    const policyId = assetHex.slice(0, 56)
+    const assetName = assetHex.slice(56)
+
+    const response = await request(
+      {
+        url: `${baseUrl}/api/txs/utxoForAddresses`,
+        method: 'post',
+        data: {addresses, asset: {policy: policyId, name: assetName}},
+      },
+      fetcherConfig,
+    )
+
+    if (isLeft(response)) {
+      handleApiError(response.error)
+    } else {
+      // @ts-ignore
+      return response.value.data[0].inline_datum.plutus_data
+    }
```

Comment:
do you have any suggestion to fix this?

<img width="562" alt="image" src="https://github.com/Emurgo/yoroi/assets/105349292/79abd6b1-6770-478f-86f0-a5157459a9a9">

Reply:
`plutus_data ` is a really complex type. Not sure how to generate the best zod schema. DO you have any suggestion?

Reply:
request takes generic